### PR TITLE
[DD4hep] DDFilteredView API Enhancement and Tracker Configuration by Era

### PR DESCRIPTION
**Pull request by ianna**
#### PR description:

* add accessor for an Iterator level: `DDFilteredView::level`
* add a `DDFilteredView::goTo` method that is used to build `GeometricDetExtras`
* add a unit test for the above
* merge DD and DD4hep `TrackerParameters` producer to allow its use in a dd4hep Era  
* introduce an Era for `trackerParameters` configuration. That allows us reusing the same RECO geometry configuration rather then its DD4hep duplication
* update geometry dictionaries for all 2021 and 2026 scenarios to allow selecting a dd4hep Era for Tracker
* regenerate all 2021 and 2026 scenarios
* simplify test configuration based on an dd4hep Era
* first step in fixing tracker numbering builder: specialisation where needed, check level to escape loops, fix configuration

The test fails in `CmsTrackerDebugNavigator` due to missing `GeometricDetExtras` (noted with a FIXME in the code).

@vargasa and @ghugo83 - FYI

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by ianna** [comment]
@cvuosalo - this PR aims to test if changes in the SpecPar names affect current Tracker geometry and workflows.

**Comment by ianna** [comment]
please test

**Comment by ghugo83** [comment]
@ianna Do you think it would be possible to have the DetectorDescription/DDCMS adapted, so that the XMLs (the SpecPars) do not need to be modified, for running with DD4hep?

I think, with the SpecPars, there are 2 issues:
- no namespace with DD4hep
- copy number referenced differently with DD4hep

**Comment by ianna** [comment]
> @ianna Do you think it would be possible to have the DetectorDescription/DDCMS adapted, so that the XMLs (the SpecPars) do not need to be modified, for running with DD4hep?

@ghugo83 it's possible to keep the SpecPars without modification in this case.

> 
> I think that, with the SpecPars with DD4hep, there are 2 issues:
> 
> * volumes in paths have no namespace with DD4hep

This is entirely up to a sub-detector developer :-)
The Ecal, for example, defines volumes with the namespaces. It's just one `std::string` now instead of two: `namespace:name`

> * copy numbers are referenced differently with DD4hep

This is also in the hands of a sub-detector developer. When a Volume is placed and a copy number is not provided, ROOT will assign it based on its own logic.

This PR removes `"//.*:"` in front of a name. The selection will work with keeping it in. However, it looks to me that the meaning of this selection is **"don't care from which file it comes from"**, e.g. drop the namespace.

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
@smuzaffar - this exception looks strange. I wouldn't think the changes in this PR could trigger it. Does it look familiar to you? Thanks.
```log
----- Begin Fatal Exception 16-Sep-2020 15:05:52 CEST-----------------------
An exception of category 'ConfigFileReadError' occurred while
   [0] Processing the python configuration file named /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-09-13-2300/src/Validation/Geometry/test/runP_Tracker.py
Exception Message:
 unknown python problem occurred.
RuntimeError: An exception of category 'PluginManagerReconfigured' occurred.


At:
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-09-13-2300/python/FWCore/PythonFramework/CmsRun.py(10): __init__
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-09-13-2300/src/Validation/Geometry/test/runP_Tracker.py(140): <module>

----- End Fatal Exception -------------------------------------------------

```

**Comment by vargasa** [comment]
@ianna thanks a lot. I was doing something similar but defined a static boolean in the abstract class which is not thread safe: https://github.com/cms-sw/cmssw/compare/master...vargasa:xTMP?expand=1. I guess that's not needed, but it's a reliever we were thinking in similar solutions :)

File: DetectorDescription/DDCMS/src/DDFilteredView.cc [hunk 1]
```diff
@@ -560,6 +560,14 @@ DDFilteredView::nav_type DDFilteredView::navPos() const {
   return pos;
 }
 
+const int DDFilteredView::level() const {
```
[the remark this answers is not in the document]
Comment (vargasa) [reply DetectorDescription/DDCMS/src/DDFilteredView.cc:563]: I see, it is faster

**Comment by ghugo83** [comment]
@ianna 
Hmm testing this with old DD, I get errors of type:

```
%MSG-e Specific:  ModuleNumbering:prod 16-Sep-2020 18:34:36 CEST Run: 1 Event: 1
Could not process q-name of a DDLogicalPart, reason:
DDLogicalPart-name "PixelForwardSensor" matching regex "PixelForwardSensor" has been found at least in following namespaces:
pixfwdbladeInnerZplus pixfwdbladeInnerZminus pixfwdbladeOuterZplus pixfwdbladeOuterZminus 
Qualify the name with a regexp for the namespace, i.e ".*:name-regexp" !
SpecPar selection is:
//PixelForwardSensor
```

**Comment by ghugo83** [comment]
>This is entirely up to a sub-detector developer :-)
The Ecal, for example, defines volumes with the namespaces. It's just one std::string now instead of two: namespace:name

Haaa ok, thanks for the info.
Which function allows access to the namespace from the DD4hep FilteredView ?

**Comment by ghugo83** [comment]
>This is also in the hands of a sub-detector developer. When a Volume is placed and a copy number is not provided, ROOT will assign it based on its own logic.

Ha yes this is annoying.
For me to understand, how are the copy numbers decided when running with old DD, when the copy numbers are not specified in XMLs? Why is it not ROOT as well? Haaa maybe in that case it is the old DD CMS interface which handles it?

File: DetectorDescription/DDCMS/src/DDFilteredView.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (ianna) [reply DetectorDescription/DDCMS/src/DDFilteredView.cc:563]: yes, I think so

**Comment by ianna** [comment]
> > This is entirely up to a sub-detector developer :-)
> > The Ecal, for example, defines volumes with the namespaces. It's just one std::string now instead of two: namespace:name
> 
> Haaa ok, thanks for the info.
> 
> You mean actually just calling name(), when defined in the subdetector, will allow namespace access?

Yes, a `Volume` (or `PlacedVolume`) `name` is the only place to store its namespace. We've discussed an option to use a `Node` `title`, but ruled it out. The latter is used now for something else (shape related)

> What needs to be done, at the detector level, to have the namespace appear? Just add the namespace in the XMLs?

When the sub-detector is built there is an option to place a `Volume` with a name. If one wants to propagate its namespace, the `PlacedVolume` name should contain it. Many algorithms strip it off explicitly by calling `noNamespase()`.

**Comment by ghugo83** [comment]
> Yes, a Volume (or PlacedVolume) name is the only place to store its namespace. We've discussed an option to use a Node title, but ruled it out. The latter is used now for something else (shape related)

Ha ok.

>When the sub-detector is built there is an option to place a Volume with a name. 

What is the name of this option? Ha in the DDDetector builder?

**Comment by ghugo83** [comment]
On this PR: why not also adding nextSibling() and parent() to the DD4hep filtered view?

**Comment by ianna** [comment]
> On this PR: why not also adding nextSibling() and parent() to the DD4hep filtered view?

These methods are implemented. However, they rely on a `SpecPar` tree structure definition.

**Comment by ianna** [comment]
@vargasa - as @ghugo83 suggested I've cleaned up the configuration. The idea is use a `dd4hep` modifier.

@cvuosalo - I'm also adding a `goTo` method which is needed to build a `GeometricDetExtra` - it is not built for DD4hep and is the reason why the navigator crashes. @vargasa - as tempting as it may seem, pleas do not use it for building the `GeometricDets` :-)

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
please test

**Comment by kpedro88** [comment]
@ianna for consistency, the same update should be propagated to 2021ZeroMaterial using the geometry script

**Comment by ianna** [comment]
> @ianna for consistency, the same update should be propagated to 2021ZeroMaterial using the geometry script

Thanks @kpedro88 ! I forgot about that. I think it would be good to generate all 2021 scenarios by default.

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
@bsunanda - you may want to update to the new API when this PR is merged. `DDFilteredView::level()` will give you the same information as `DDFilteredView::copyNos().size()`, but it is much faster.

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
> @cvuosalo - this PR aims to test if changes in the SpecPar names affect current Tracker geometry and workflows.

Ok, the xml changes reverted.

**Comment by ianna** [comment]
> This leads to
> 
> ```
> %MSG-e CmsTrackerLayerBuilder:  ModuleNumbering:prod 18-Sep-2020 13:17:32 CEST  Run: 1 Event: 1
>  ERROR - I was expecting a String, Rod or Ladder, I got a PixelBarrelDet
> %MSG
> ```
> 
> and
> 
> ```
> %MSG-e CmsTrackerSubStrctBuilder:  ModuleNumbering:prod  18-Sep-2020 13:17:33 CEST Run: 1 Event: 1
>  ERROR - I was expecting a Layer ,Wheel or Disk... I got a PixelBarrelDet
> %MSG
> ```
> 
> when navigating with DD4hep.
> 
> But the level() in the API is quite useful, I use it in my solution.
> Maybe merge this first, and I will push the fix of the DD4hep navigation (plus issues down the line in the numbering) on top of this PR.
> 
> I get no more navigation error of this type, no crash in navigation, and basically the exact same navigation with DD4hep as with old DD.

Thanks @ghugo83! I was mostly working on the configuration infrastructure and the unit tests. The changes in `TrackerNumbering` are meant as an explanation what the `level()` does :-)

@cvuosalo and @kpedro88 - if introducing a dd4hep Era brings no problems to the 2026 workflows, I'd like you to merge the PR asap. Then we can branch our development more efficiently.

@bsunanda may want to use the `level()` method and the `goTo() `method is essential for a `GeometricDetExtra` builder.

**Comment by ianna** [comment]
urgent

**Comment by kpedro88** [comment]
@ianna this PR will now conflict with #31451 that introduces a new Phase 2 geometry

**Comment by ianna** [comment]
> @ianna this PR will now conflict with #31451 that introduces a new Phase 2 geometry

I'd think one of the PRs goes in first and the other should be rebased. What's an ETA of #31451?

**Comment by ghugo83** [comment]
@ianna Now that the navigation is fixed (need to be cleaned, but defo works), I am down the line fixing the DetId numbering with DD4hep.

First I must say that there was a huge number of DetId computation regressions with DD4hep (outside of no operational navigation, which is fine now).
I have fixed most of the regressions in DetId computation with DD4hep.
Still few issues, due to unsafe handling of floating point values in the legacy code : hence the different rounding of positions / phi representation with DD4hep, sometimes lead to different sorting, hence ids. 
Should be done in the afternoon with them.

There are 2 points I am wondering about:

- Is there a way to speed up firstChild() in the DD4hep FilteredView interface?

- At https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerNumberingBuilder/plugins/DDDCmsTrackerContruction.cc#L54 
The filtered view is only filtered with the "TkDDDStructure" attribute.
With DD4hep, this leads to https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerNumberingBuilder/plugins/CmsDetConstruction.cc#L16 being never true, because it asks for the "TrackerStereoDetectors" attribute, and the view has already been filtered with TkDDDStructure".
Hence, no stereo sensors are detected, and hence a shift of 1 in the Ids.
Basically, this comes from the DDCMS FilteredView interface having a different behavior than the one in old DD.
Is there a way to re-build a filtered view, which has already been filtered, with a "TrackerStereoDetectors" filter (and other filters, there are also ROCNumber filters, etc)?
Or one would need to compute a different filtered view in https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerNumberingBuilder/plugins/DDDCmsTrackerContruction.cc#L54 for each filter of interest, and pass it as argument all the way till GeometricDet (sounds bad)?


NB: Please do not modify this PR to avoid geant merge conflicts, I am working on a base on top of it ;p

**Comment by ianna** [comment]
@kpedro88 and @cvuosalo - could we, please, converge before the weekend?

> @ianna Now that the navigation is fixed (need to be cleaned, but defo works), I am down the line fixing the DetId numbering with DD4hep.
> 
> First, I must say that there was a huge number of DetId computation regressions with DD4hep (outside of no operational navigation, which is fine now).
> I have fixed most of the regressions in DetId computation with DD4hep.
> Still few issues, due to unsafe handling of floating point values in the legacy code : hence the different rounding of positions / phi representation with DD4hep, sometimes lead to different sorting, hence ids.
> Should be done in the afternoon with them.
> 
> There are 2 points I am wondering about:
> 
> * Is there a way to speed up firstChild() in the DD4hep FilteredView interface?
> * At https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerNumberingBuilder/plugins/DDDCmsTrackerContruction.cc#L54
>   The filtered view is only filtered with the "TkDDDStructure" attribute.
>   With DD4hep, this leads to https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerNumberingBuilder/plugins/CmsDetConstruction.cc#L16 being never true, because it asks for the "TrackerStereoDetectors" filter, and the view has already been filtered with TkDDDStructure".
>   Hence, no stereo sensors are detected, and hence a shift of 1 in the Ids.
>   Basically, this comes from the DDCMS FilteredView interface having a different behavior than the one in old DD.
>   Is there a way to re-build a filtered view, which has already been filtered by "TkDDDStructure", with a "TrackerStereoDetectors" filter (and other filters, there are also ROCNumber filters, etc)?
>   Or one would need to compute a different filtered view, from the compact view in https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerNumberingBuilder/plugins/DDDCmsTrackerContruction.cc#L54 , for each filter of interest, and pass it as argument all the way till GeometricDet (sounds bad)?
> 
> NB: Please do not modify this PR to avoid geant merge conflicts, I am working on a base on top of it ;p

Excellent! 

Yes, the navigation definitely can be faster. I'll open a new PR ~soonish with a unit test showing how fast it can be. We can discuss it next week. The main problem, I think, is that we need all the regression tests automatised to make sure we do not "throw the baby out with the bathwater" :-)

I'm freezing this PR. Thanks again!

**Comment by ghugo83** [comment]
@ianna @vargasa Another point: had you guys tried to check the Tracker XMLs parsing with DD4hep (constant sections evaluation, copy numbers, units issues)?
Is it ok for all Tracker scenarios?

When creating DDDetector from Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D65.xml (instead of 2021), the XML parsing crashes. And I suspect undetected constant evaluations issues in the 2021 XMLs when running with DD4hep (as I faced with COCOA and PPS ports to DD4hep).
I guess there should be efforts here as well (the files to be touched are XMLs, and independent from the TrackerNumbering code work), so it can be done in parallel.

**Comment by ianna** [comment]
> @ianna @vargasa Another point: had you guys tried to check the Tracker XMLs parsing with DD4hep (constant sections evaluation, copy numbers, units issues)?
> Is it ok for all Tracker scenarios?
> 
> When creating DDDetector from Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D65.xml (instead of 2021), the XML parsing crashes.

Yes, I've done it at some point. D65 was not there yet :-(
I can have a look what's wrong with it.

> Additionally, I suspect undetected constant evaluations issues in the 2021 XMLs themselves, when running with DD4hep (as I faced with COCOA and PPS ports to DD4hep).

A debug section if added to the xml can help to trace it:
```
<debug>
    <debug_shapes/>
    <debug_includes/>
    <debug_rotations/>
    <debug_includes/>
    <debug_volumes/>
    <debug_constants/>
    <debug_namespaces/>
    <debug_placements/>
    <debug_algorithms/>
    <debug_materials/>
    <debug_visattr/>
 </debug>
```
Please, see https://cmssdt.cern.ch/lxr/source/DetectorDescription/DDCMS/data/cms-test-ddangular-algorithm.xml

> I guess there should be efforts in the XMLs as well. This is relatively independent from the TrackerNumbering code work, so it can be done in parallel.

Yes, I agree. We need automated regression test for all scenarios.

**Comment by ianna** [comment]
@ghugo83 - Hmm... 2026D65 looks fine to me. Please, check my area on `cmsdev20`:
```
/build/yana/DD4HEP/CMSSW_11_2_X_2020-09-13-2300/src $ cmsRun DetectorDescription/DDCMS/test/python/testGeometry2026D65.py 
```
Note, I've modified the scenario to print the debug statements:
```diff
$ git diff
diff --git a/Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D65.xml b/Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D65.xml
index 9500682ca1e..ba17afc98cc 100644
--- a/Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D65.xml
+++ b/Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2026D65.xml
@@ -1,5 +1,18 @@
 <?xml version="1.0"?>
 <DDDefinition>
+  <debug>
+    <debug_shapes/>
+    <debug_includes/>
+    <debug_rotations/>
+    <debug_includes/>
+    <debug_volumes/>
+    <debug_constants/>
+    <debug_namespaces/>
+    <debug_placements/>
+    <debug_algorithms/>
+    <debug_materials/>
+    <debug_visattr/>
+ </debug>
   <open_geometry/>
   <close_geometry/>

```

**Comment by ianna** [comment]
+1

@cvuosalo - please, put it on hold it you object. Thanks.

**Comment by ianna** [comment]
@kpedro88 - please, let me know what would you like me to do with this PR. Thanks.

**Comment by kpedro88** [comment]
@ianna I prefer to let #31451 be merged first and then this PR can propagate the update to D66.

**Comment by vargasa** [comment]
@ghugo83 those are great news! please let me know how to reproduce the issues you saw

File: Geometry/TrackerGeometryBuilder/python/trackerParameters_cff.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import FWCore.ParameterSet.Config as cms
+
+from Geometry.TrackerGeometryBuilder.trackerParameters_cfi import *
+
+from Configuration.ProcessModifiers.dd4hep_cff import dd4hep
+
+dd4hep.toModify(trackerParameters,
+                fromDD4Hep = cms.bool(True),
```
Comment (cvuosalo) [inline Geometry/TrackerGeometryBuilder/python/trackerParameters_cff.py:8]: Does this line enable DD4hep Tracker parameters for all Extended 2021 workflows? We are not ready for that yet.

Comment (ianna) [reply Geometry/TrackerGeometryBuilder/python/trackerParameters_cff.py:8]: No, only when a dd4hep Era is selected

File: Geometry/TrackerGeometryBuilder/test/trackerParametersFromDD4hep_cfg.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+process.source = cms.Source("EmptySource")
+
+process.maxEvents = cms.untracked.PSet(
+    input = cms.untracked.int32(1)
+)
+
+process.test = cms.EDAnalyzer("TrackerParametersAnalyzer")
+
+process.p1 = cms.Path(process.test)
+
```
Comment (cvuosalo) [inline Geometry/TrackerGeometryBuilder/test/trackerParametersFromDD4hep_cfg.py:18]: What part of this config is DD4hep? GeometryExtended is Run 1. Is the name of the file incorrect?

**Comment by bsunanda** [comment]
@ianna what does level() and goto() method provide? Is level() provides # of levels which we are getting from copyNos().size()? I shall wait for this PR to merge to modify Muon and Calorimeter codes

**Comment by ianna** [comment]
> @ianna what does level() and goto() method provide? Is level() provides # of levels which we are getting from copyNos().size()? I shall wait for this PR to merge to modify Muon and Calorimeter codes

Yes, `level()` provides # of levels which we are getting from `copyNos().size()`

File: Geometry/TrackerGeometryBuilder/test/trackerParametersFromDD4hep_cfg.py (hunk 3 above)
Comment (ianna) [reply Geometry/TrackerGeometryBuilder/test/trackerParametersFromDD4hep_cfg.py:18]: No, the configuration should and has been changed locally. I've frozen this PR on request from @ghugo83 .

**Comment by ianna** [comment]
> @ianna this PR will now conflict with #31451 that introduces a new Phase 2 geometry

@kpedro88 - the PR mentioned has been merged. I've updated all the scenarios in #31517. Please, let me know if you prefer this PR to be merged or the #31517?

**Comment by ianna** [comment]
please test

**Comment by ghugo83** [comment]
Yayyyy after additional fixes in numbering, i get a 1:1 comparison between old DD and DD4hep, for not only navigation, but DetIds now!
Need to clean things and commits into sth which is PR-able, but I have finished spotting all regressions sources I think (there were around 10 of them ;p).

And it does not seem to have issue in XMLs, at least for 2021 actually. 
I was getting different Phi with DD4hep than old DD, for same modules. But in the end, it was just because module.Rho() = 0, and DD4hep has a different default behavior on this.

**Comment by ghugo83** [comment]
Sorry it is a bit off-topic of this PR though.

**Comment by ianna** [comment]
> Yayyyy after additional fixes in numbering, i get a 1:1 comparison between old DD and DD4hep, for not only navigation, but DetIds now!
> Need to clean things and commits into sth which is PR-able, but I have finished spotting all regressions sources I think (there were around 10 of them ;p).
> 
> And it does not seem to have issue in XMLs, at least for 2021 actually.
> I was getting different Phi with DD4hep than old DD, for same modules. But in the end, it was just because module.Rho() = 0, and DD4hep has a different default behavior on this.

Excellent! I think, I can close this PR then :-)

**Comment by ghugo83** [comment]
>Excellent! I think, I can close this PR then :-) 

@ianna Hmm why closing this PR? The API enhancement is nice anyway.
I can PR on top of this PR?

**Comment by bsunanda** [comment]
I was waiting this PR to be merged to make the new scenario with muons. Why do you close this PR?
________________________________

**Comment by ianna** [comment]
> > Excellent! I think, I can close this PR then :-)
> 
> @ianna Hmm why closing this PR? The API enhancement is nice anyway.
> I can PR on top of this PR?

@ghugo83 - the API has been merged already. The dd4hep era introduction has been moved to a new PR based on a fresh master. When you'll make a PR, it will cover the rest :-)

**Comment by ianna** [comment]
> I was waiting this PR to be merged to make the new scenario with muons. Why do you close this PR?
> […](#)

I hope, https://github.com/cms-sw/cmssw/pull/31517 covers it

**Comment by ianna** [comment]
ok, keep the options open ;-)

**Comment by ghugo83** [comment]
@ianna Could you please solve conflicts and merge? :) 
So that I can open my PR which fixes Tracker navigation and DetIds with DD4hep (still need to look whether evth else in sim step is ok, so there is a fair bit to look at, but I expect not too many regressions downstream now that the 'upstream' is fixed).

**Comment by ianna** [comment]
> @ianna Could you please solve conflicts and merge? :)
> So that I can open my PR which fixes Tracker navigation and DetIds with DD4hep (still need to look whether evth else in sim step is ok, so there is a fair bit to look at, but I expect not too many regressions downstream now that the 'upstream' is fixed).

Sorry for the wait - I'm on it :-)

**Comment by silviodonato** [comment]
please test

**Comment by kpedro88** [comment]
@ianna this PR needs to be rebased now that #31517 is merged

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
+1

**Comment by ghugo83** [comment]
@iana Ha sorry, did not see that you introduced level() is another PR, which got merged.
So you can safely close this now (navigation is not working here, and this conflicts with my branch where I fix navigation).

**Comment by ianna** [comment]
> @Iana Ha sorry, did not see that you introduced level() is another PR, which got merged.
> So you can safely close this now (navigation is not working here, and this conflicts with my branch where I fix navigation).

@ghugo83 - yes, this was my intention and if it makes easier for you I'll close this one :-)
